src/init.cpp

```diff
@@ -1348,10 +1347,23 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
             const int64_t load_block_index_start_time = GetTimeMillis();
             try {
                 LOCK(cs_main);
-                chainman.InitializeChainstate(Assert(node.mempool.get()));
                 chainman.m_total_coinstip_cache = nCoinCacheUsage;
                 chainman.m_total_coinsdb_cache = nCoinDBCache;
 
+                // Load a chain created from a UTXO snapshot, if any exist.
+                chainman.DetectSnapshotChainstate(*Assert(node.mempool));
+
+                // Conservative value that will ultimately be changed by
+                // a call to `chainman.MaybeRebalanceCaches()`.
+                double init_cache_fraction = 0.2;
+
+                // If we're not using a snapshot or we haven't fully validated it yet,
+                // create a validation chainstate.
+                if (!chainman.IsSnapshotValidated()) {
+                    LogPrintf("Loading validation chainstate\n");
+                    chainman.InitializeChainstate(Assert(node.mempool.get()));
+                }
```

Comment:
I took a run at doing this (https://github.com/jamesob/bitcoin/commit/cec687abb4f587d756bbaac62c02c09704cd8a42), but the refactor necessary to encapsulate all of the chainstate init logic is fairly complicated and difficult to verify correctness for. I burned more time than I'd like to admit trying to get this to work, but was getting a weird race/segfault with the loadblk thread as a result of moving this stuff around. So rather than further bloat this changeset I'm going to leave this for a future PR.